ethcore/src/snapshot/mod.rs

```diff
@@ -0,0 +1,419 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Snapshot creation helpers.
+
+use std::collections::VecDeque;
+use std::fs::{create_dir_all, File};
+use std::io::Write;
+use std::path::{Path, PathBuf};
+
+use account_db::{AccountDB, AccountDBMut};
+use client::BlockChainClient;
+use error::Error;
+use ids::BlockID;
+use views::{BlockView, HeaderView};
+
+use util::{Bytes, Hashable, HashDB, JournalDB, snappy, TrieDB, TrieDBMut, TrieMut};
+use util::hash::{FixedHash, H256};
+use util::rlp::{DecoderError, RlpStream, Stream, UntrustedRlp, View};
+
+use self::account::Account;
+use self::block::AbridgedBlock;
+
+use crossbeam::{scope, ScopedJoinHandle};
+
+mod account;
+mod block;
+
+// Try to have chunks be around 16MB (before compression)
+const PREFERRED_CHUNK_SIZE: usize = 16 * 1024 * 1024;
+
+/// Take a snapshot using the given client and database, writing into `path`.
+pub fn take_snapshot(client: &BlockChainClient, mut path: PathBuf, state_db: &HashDB) -> Result<(), Error> {
+	let chain_info = client.chain_info();
+
+	let genesis_hash = chain_info.genesis_hash;
+	let best_header_raw = client.best_block_header();
+	let best_header = HeaderView::new(&best_header_raw);
+	let state_root = best_header.state_root();
+
+	trace!(target: "snapshot", "Taking snapshot starting at block {}", best_header.number());
+
+	let _ = create_dir_all(&path);
+
+	let state_hashes = try!(chunk_state(state_db, &state_root, &path));
+	let block_hashes = try!(chunk_blocks(client, best_header.hash(), genesis_hash, &path));
+
+	trace!(target: "snapshot", "produced {} state chunks and {} block chunks.", state_hashes.len(), block_hashes.len());
+
+	let manifest_data = ManifestData {
+		state_hashes: state_hashes,
+		block_hashes: block_hashes,
+		state_root: state_root,
+		block_number: chain_info.best_block_number,
+		block_hash: chain_info.best_block_hash,
+	};
+
+	path.push("MANIFEST");
+
+	let mut manifest_file = try!(File::create(&path));
+
+	try!(manifest_file.write_all(&manifest_data.to_rlp()));
+
+	Ok(())
+}
+
+// shared portion of write_chunk
+// returns either a (hash, compressed_size) pair or an io error.
+fn write_chunk(raw_data: &[u8], compression_buffer: &mut Vec<u8>, path: &Path) -> Result<(H256, usize), Error> {
+	let compressed_size = snappy::compress_into(raw_data, compression_buffer);
+	let compressed = &compression_buffer[..compressed_size];
+	let hash = compressed.sha3();
+
+	assert!(snappy::validate_compressed_buffer(compressed));
```

Comment:
this assert should be removed outright actually. i added it while debugging but it never fails